scvi/model/_jaxscvi.py

```diff
@@ -0,0 +1,342 @@
+
+    @property
+    def module(self):
+        if self._module is None:
+            self._module = self._get_module()
+        return self._module
+
+    def train(
+        self,
+        max_epochs: Optional[int] = None,
+        use_gpu: Optional[Union[str, int, bool]] = None,
+        train_size: float = 0.9,
+        validation_size: Optional[float] = None,
+        batch_size: int = 128,
+        lr: float = 1e-3,
+    ):
+        """
+        Train the model.
+
+        Parameters
+        ----------
+        max_epochs
+            Number of passes through the dataset. If `None`, defaults to
+            `np.min([round((20000 / n_cells) * 400), 400])`
+        use_gpu
+            Use default GPU if available (if None or True), or index of GPU to use (if int),
+            or name of GPU (if str, e.g., `'cuda:0'`), or use CPU (if False).
+        train_size
+            Size of training set in the range [0.0, 1.0].
+        validation_size
+            Size of the test set. If `None`, defaults to 1 - `train_size`. If
+            `train_size + validation_size < 1`, the remaining cells belong to a test set.
+        batch_size
+            Minibatch size to use during training.
+        """
+        if max_epochs is None:
+            n_cells = self.adata.n_obs
+            max_epochs = np.min([round((20000 / n_cells) * 400), 400])
+
+        data_splitter = DataSplitter(
+            self.adata_manager,
+            train_size=train_size,
+            validation_size=validation_size,
+            batch_size=batch_size,
+            # for pinning memory only
+            use_gpu=False,
+            iter_ndarray=True,
+        )
+        data_splitter.setup()
+        train_loader = data_splitter.train_dataloader()
+
+        module_kwargs = self.module_kwargs.copy()
+        module_kwargs.update(dict(is_training=True))
+        module = self._get_module(module_kwargs)
+
+        # if key is generated on CPU, model params will be on CPU
+        # we have to pay the price of a JIT compilation though
+        if use_gpu is False:
+            key = jax.jit(lambda i: random.PRNGKey(i), backend="cpu")
+        else:
+            # dummy function
+            def key(i: int):
+                return random.PRNGKey(i)
+
+        self.rngs = {
+            "params": key(0),
+            "dropout": key(1),
+            "z": key(2),
+        }
+        module_init = module.init(self.rngs, next(iter(train_loader)))
+        params = module_init["params"]
+        batch_stats = module_init["batch_stats"]
+
+        state = TrainState.create(
+            apply_fn=module.apply,
+            params=params,
+            tx=optax.adamw(lr, eps=0.01, weight_decay=1e-6),
+            batch_stats=batch_stats,
+        )
+
+        @jax.jit
+        def train_step(state, array_dict, rngs, kl_weight=1) -> jnp.ndarray:
+            rngs = {k: random.split(v)[1] for k, v in rngs.items()}
+
+            # batch stats can't be passed here
```

Comment:
yes

Reply:
should be careful bc if the inputs are the same but batch_stats is changed (probably wont happen but still), it might return the cached output. is this the flax recommended way?

Reply:
this is what I found in flax's examples